pkg/query/caveat.go

```diff
@@ -0,0 +1,309 @@
+package query
+
+import (
+	"fmt"
+
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+// CaveatIterator wraps another iterator and applies caveat evaluation to its results.
+// It checks caveat conditions on relationships during iteration and only yields
+// relationships that satisfy the caveat constraints.
+type CaveatIterator struct {
+	subiterator Iterator
+	caveat      *core.ContextualizedCaveat
+}
+
+var _ Iterator = &CaveatIterator{}
+
+// NewCaveatIterator creates a new caveat iterator that wraps the given subiterator
+// and applies the specified caveat conditions.
+func NewCaveatIterator(subiterator Iterator, caveat *core.ContextualizedCaveat) *CaveatIterator {
+	return &CaveatIterator{
+		subiterator: subiterator,
+		caveat:      caveat,
+	}
+}
+
+func (c *CaveatIterator) CheckImpl(ctx *Context, resources []Object, subject ObjectAndRelation) (PathSeq, error) {
```

Comment:
The implementation of `CheckImpl`, `IterSubjectsImpl` and `IterResourcesImp` look to be exactly the same, except for the very first line. Can we remove the duplication?